### DISCUSSION TITLE
Repository has unknown error: Pushing to fatal: could not read Username for: No such device or address

### DISCUSSION THREAD

**Discussion by deckstose** [renovatebot/renovate#29165] [Request Help]
### What would you like help with?

I would like help with my configuration

### How are you running Renovate?

Self-hosted

### If you're self-hosting Renovate, tell us which platform (GitHub, GitLab, etc) and which version of Renovate.

Docker, 37

### Please tell us more about your question or problem

When Renovate tries to push a branch to my repository, it says it can't read the username for my domain.

```
message": "Pushing to http://domain/user/repo.git\nfatal: could not read Username for 'https://domain': No such device or address\n
```

I'm running the docker image for a self-hosted Gitea instance. It detects the repository but then isn't able to push to it. I thought that the Gitea token should be enough to create new branches, but I also added `hostRules` with the username and password which doesn't fix it. The `renovate` user on Gitea has write access to the repository in question. 

I had a lot of troubles with getting Renovate to resolve domains from my custom DNS, so for testing purposes I hard coded in an entry for my Gitea instance in the containers `/etc/hosts`. Renovate now resolves the address.

I also tried this 
```bash
docker exec renovate git clone https://domain/user/repo.git /tmp/renovate/repo
```
which also works. Might not be a DNS problem.

The repository is otherwise clean, just the master branch and a single commit.

This is my `config.js`:

```
module.exports = {
	endpoint: 'https://domain',
	token: '**********',
	platform: 'gitea',
	gitAuthor: 'Renovate Bot <test@mail.xyz>',
	autodiscover: true,
	autodiscoverFilter: ['user/*'],
	onboarding: true,
	onboardingConfig: {
		extends: ['config:recommended'],
	},
	binarySource: 'docker',
	persistRepoData: true,
	hostRules: [
		{
			matchHost: 'domain',
			username: 'renovate',
			password: 'password',
		},
		{
			matchHost: 'http://domain',
			username: 'renovate',
			password: 'password',
		},
		{
			matchHost: 'https://domain',
			username: 'renovate',
			password: 'password',
		},
	],
};
```

I find that the docs split things up a lot (at least for what I want to do), so I had to copy configs from here and there and I have no real idea what most of these options do together.

### Logs (if relevant)

<details><summary>Logs</summary>

```
renovate  | DEBUG: onboarding config (repository=user/repo)
renovate  |        "config": {
renovate  |          "$schema": "https://docs.renovatebot.com/renovate-schema.json",
renovate  |          "extends": ["config:recommended"]
renovate  |        }
renovate  | DEBUG: Creating onboarding branch (repository=user/repo)
renovate  | DEBUG: Preparing files for committing to branch renovate/configure (repository=user/repo)
renovate  | DEBUG: Setting git author name: Renovate Bot (repository=user/repo)
renovate  | DEBUG: Setting git author email: test@mail.xyz (repository=user/repo)
renovate  | DEBUG: git commit (repository=user/repo)
renovate  |        "deletedFiles": [],
renovate  |        "ignoredFiles": [],
renovate  |        "result": {
renovate  |          "author": null,
renovate  |          "branch": "renovate/configure",
renovate  |          "commit": "e1fb7d9be04cb38965c2c58b1810ef18fb8d7e98",
renovate  |          "root": false,
renovate  |          "summary": {"changes": 1, "insertions": 6, "deletions": 0}
renovate  |        }
renovate  | DEBUG: Pushing refSpec renovate/configure:renovate/configure (repository=user/repo)
renovate  | DEBUG: Git function thrown (repository=user/repo)
renovate  |        "err": {
renovate  |          "task": {
renovate  |            "commands": [
renovate  |              "push",
renovate  |              "origin",
renovate  |              "renovate/configure:renovate/configure",
renovate  |              "--force-with-lease",
renovate  |              "-u",
renovate  |              "--no-verify",
renovate  |              "--verbose",
renovate  |              "--porcelain"
renovate  |            ],
renovate  |            "format": "utf-8",
renovate  |            "parser": "[function]"
renovate  |          },
renovate  |          "message": "Pushing to http://domain/user/repo.git\nfatal: could not read Username for 'https://domain': No such device or address\n",
renovate  |          "stack": "Error: Pushing to http://domain/user/repo.git\nfatal: could not read Username for 'https://**redacted**@3.24.0/node_modules/simple-git/src/lib/plugins/error-detection.plugin.ts:42:29)\n    at PluginStore.exec (/usr/local/renovate/node_modules/.pnpm/simple-git@3.24.0/node_modules/simple-git/src/lib/plugins/plugin-store.ts:54:29)\n    at /usr/local/renovate/node_modules/.pnpm/simple-git@3.24.0/node_modules/simple-git/src/lib/runners/git-executor-chain.ts:124:42\n    at new Promise (<anonymous>)\n    at GitExecutorChain.handleTaskData (/usr/local/renovate/node_modules/.pnpm/simple-git@3.24.0/node_modules/simple-git/src/lib/runners/git-executor-chain.ts:121:14)\n    at GitExecutorChain.<anonymous> (/usr/local/renovate/node_modules/.pnpm/simple-git@3.24.0/node_modules/simple-git/src/lib/runners/git-executor-chain.ts:97:40)\n    at Generator.next (<anonymous>)\n    at fulfilled (/usr/local/renovate/node_modules/.pnpm/simple-git@3.24.0/node_modules/simple-git/dist/cjs/index.js:52:24)\n    at processTicksAndRejections (node:internal/process/task_queues:95:5)"
renovate  |        }

```

</details>

**Comment by rarkins** [comment]
You'll need to find an answer by reading through https://www.google.com/search?q=git+%22could+not+read+username%22+push+%22No+such+device+or+address%22

If you discover that Renovate needs to change something, please let us know. But most likely this sounds local to your environment

**Comment by deckstose** [comment]
So you mean I need to configure ssh? The instance is set up to allow pushing via http/https.

Should renovate usually be able to push just with the access token?

**Reply by deckstose** [reply]
So the error comes from git not being in an interactive environment and not being able to figure out what username to use. From [this discussion](https://github.com/orgs/community/discussions/26580#discussioncomment-7147217) I tried to use `https://<token>@<domain>/<repo>/<user>.git` and this works fine. I'd expect Renovate to use this style of remote url to push when using http/https. I wonder how this works for other folks and why no one else has had issues like that.

Is it possible that the configured `hostRules` are not applied, since they specify a username which is not used.

**Reply by rarkins** [reply]
I only meant, please read the Google results until you find one which seems to relate to your case which solves the problem for you

**Reply by deckstose** [reply]
Yes, I'm certain that I know the problem, look what I stated above.

**Comment by viceice** [comment]
which version of gitea are you using?

**Reply by deckstose** [reply]
1.21.11 inside docker

**Reply by deckstose** [reply]
I just tested it with a token that has write permission for all categories and it works now. So it is a token issue.

**Comment by deckstose** [comment]
This had nothing todo with renovate but rather with dns. If you have a dns server inside of a docker container other containers on the same host are not able to resolve addresses with this container.
